File: src/acemusic/song_structure.py
"""Song-structure planning for the full-song auto-extend pipeline (US-6.7).

Given a seed clip duration and a target total length, produces an ordered list
of sections describing how to grow the clip into a complete song. Pure
functions only: no I/O, no audio processing.
"""

from __future__ import annotations

from dataclasses import dataclass

SONG_STRUCTURE: list[str] = [
    "intro",
    "verse",
    "chorus",
    "verse",
    "chorus",
    "bridge",
    "outro",
]

# Per-section (relative_weight, style_hint).
# Weights are normalized across the full SONG_STRUCTURE list; each occurrence
# (e.g. the two verses, two choruses) contributes its weight independently.
SECTION_CONFIG: dict[str, tuple[float, str]] = {
    "intro": (1.0, "intro, atmospheric build, sparse arrangement"),
    "verse": (2.0, "verse, melodic, narrative groove"),
    "chorus": (2.0, "chorus, hook, energetic, full arrangement"),
    "bridge": (1.5, "bridge, transition, contrast and tension"),
    "outro": (1.0, "outro, fade and resolve"),
}

MIN_SECTION_SECONDS: float = 4.0


@dataclass(frozen=True)
class Section:
    """A planned section: what to generate, how long, with what style emphasis."""

    name: str
    duration_s: float
    style_hint: str


def plan_sections(
    seed_duration: float,
    target_duration: int | float,
    structure: list[str] | None = None,
) -> list[Section]:
    """Distribute (target_duration - seed_duration) across an ordered section list.

    ``structure`` defaults to the seven canonical sections (:data:`SONG_STRUCTURE`)
    but may be overridden — e.g. the full-song API's ``structure_plan`` (US-10.4).
    Every name must have a :data:`SECTION_CONFIG` entry. Returns sections in the
    given order; each receives a slice of the remaining time proportional to its
    configured weight.

    When there is enough headroom, each section is rounded up to
    MIN_SECTION_SECONDS so it produces an audible chunk. When the remainder is
    too small to honor the minimum across all sections, the floor is dropped
    and durations stay proportional to weights — so the planned total never
    overshoots ``target_duration``. Every section still gets a positive
    duration.

    Raises ValueError if ``structure`` is empty or names an unknown section, or
    if target_duration is non-positive or shorter than the seed.
    """
    sections = SONG_STRUCTURE if structure is None else list(structure)
    if not sections:
        raise ValueError("structure must contain at least one section")
    unknown = [name for name in sections if name not in SECTION_CONFIG]
    if unknown:
        raise ValueError(f"unknown section(s) {unknown}; valid sections are {sorted(SECTION_CONFIG)}")
    if target_duration <= 0:
        raise ValueError(f"target_duration must be positive, got {target_duration}")
    if target_duration <= seed_duration:
        raise ValueError(f"target_duration ({target_duration}s) must exceed seed_duration ({seed_duration}s)")

    remaining = float(target_duration) - float(seed_duration)
    weights = [SECTION_CONFIG[name][0] for name in sections]
    total_weight = sum(weights)
    raw_durations = [remaining * (w / total_weight) for w in weights]

    # Apply the audible-section floor only if doing so still fits inside the
    # remaining budget. At moderate margins (e.g. seed=30, target=60) the floor
    # would bump small sections up enough to overshoot remaining, so fall back
    # to the raw proportional distribution in that case.
    floored = [max(d, MIN_SECTION_SECONDS) for d in raw_durations]
    durations = floored if sum(floored) <= remaining else raw_durations

    return [
        Section(name=name, duration_s=duration, style_hint=SECTION_CONFIG[name][1])
        for name, duration in zip(sections, durations)
    ]

File: web/lib/song-structure.ts
// Song-structure planning for the Get Full Song flow (US-17.4).
//
// A TypeScript port of the backend planner (src/acemusic/song_structure.py,
// plan_sections) so the wizard can preview the section breakdown and size each
// extend without an extra API round-trip. The structure and per-section weights
// are static, spec-defined data (spec §21); keep the values here in lockstep
// with SECTION_CONFIG on the backend. Pure functions only.

import type { Clip } from "@/lib/workspace-clips"
import { DURATION_MAX } from "@/lib/constants/generation"

/** The canonical seven-section song shape (intro → outro). */
export const SONG_STRUCTURE = [
  "intro",
  "verse",
  "chorus",
  "verse",
  "chorus",
  "bridge",
  "outro",
] as const

export type SectionName = (typeof SONG_STRUCTURE)[number]

/**
 * Per-section `[relativeWeight, styleHint]`, normalized across the whole
 * structure — each occurrence (both verses, both choruses) contributes its
 * weight independently. Mirrors the backend SECTION_CONFIG exactly.
 */
export const SECTION_CONFIG: Record<SectionName, readonly [number, string]> = {
  intro: [1.0, "intro, atmospheric build, sparse arrangement"],
  verse: [2.0, "verse, melodic, narrative groove"],
  chorus: [2.0, "chorus, hook, energetic, full arrangement"],
  bridge: [1.5, "bridge, transition, contrast and tension"],
  outro: [1.0, "outro, fade and resolve"],
}

/** Round each section up to this many seconds when the budget allows (audible chunk). */
export const MIN_SECTION_SECONDS = 4

/** Longest seed a full-song build accepts; also the clip-card eligibility cutoff. */
export const MAX_SEED_DURATION = 60
/** Target-duration slider bounds (seconds). Max tracks the backend generation cap. */
export const TARGET_DURATION_MIN = 120
export const TARGET_DURATION_MAX = DURATION_MAX // 240
/** Default target when the wizard opens. */
export const DEFAULT_TARGET_DURATION = 210

/** A planned section: what to generate, how long, with what style emphasis. */
export type Section = {
  name: SectionName
  durationSeconds: number
  styleHint: string
}

/**
 * Distribute `(targetDuration - seedDuration)` across the canonical structure,
 * each section getting a slice proportional to its configured weight.
 *
 * When there is headroom every section is rounded up to MIN_SECTION_SECONDS so
 * it produces an audible chunk; when the remainder is too small to honor that
 * floor across all sections the floor is dropped and durations stay purely
 * proportional — so the planned total never overshoots `targetDuration`. Every
 * section still gets a positive duration. Mirrors the backend `plan_sections`.
 *
 * Returns `[]` for a non-positive or too-short target rather than throwing —
 * the caller clamps the slider to a valid range, and an empty plan reads as
 * "nothing to do" in the UI.
 */
export function planSections(
  seedDuration: number,
  targetDuration: number
): Section[] {
  const remaining = targetDuration - seedDuration
  if (remaining <= 0) return []

  const weights = SONG_STRUCTURE.map((name) => SECTION_CONFIG[name][0])
  const totalWeight = weights.reduce((a, b) => a + b, 0)
  const raw = weights.map((w) => remaining * (w / totalWeight))

  // Apply the audible-section floor only if doing so still fits the budget;
  // otherwise fall back to the raw proportional split (never overshoots).
  const floored = raw.map((d) => Math.max(d, MIN_SECTION_SECONDS))
  const flooredTotal = floored.reduce((a, b) => a + b, 0)
  const durations = flooredTotal <= remaining ? floored : raw

  return SONG_STRUCTURE.map((name, i) => ({
    name,
    durationSeconds: durations[i],
    styleHint: SECTION_CONFIG[name][1],
  }))
}

/**
 * The whole-second duration actually requested when extending for a section.
 * Floored (not rounded) so summed sections never push the cumulative clip past
 * the backend's generation cap; every section still gets at least one second.
 * Shared by the generation step (what to send) and the completion summary (what
 * was assembled).
 */
export function sectionExtendSeconds(section: Section): number {
  return Math.max(1, Math.floor(section.durationSeconds))
}

/**
 * A clip is eligible for the Full Song flow when it has a known duration under
 * MAX_SEED_DURATION — the flow grows a short seed into a full song, so it only
 * makes sense for short clips.
 */
export function isFullSongEligible(clip: Pick<Clip, "duration">): boolean {
  return clip.duration != null && clip.duration < MAX_SEED_DURATION
}
